ql/ql/src/queries/overlay/InlineOverlayCaller.ql

```diff
@@ -0,0 +1,41 @@
+/**
+ * @name Cannot inline predicate across overlay frontier
+ * @description Local inline predicates that are not annotated with `overlay[caller]` are
+ *              not inlined across the overlay frontier. This may negatively affect performance.
+ * @kind problem
+ * @problem.severity warning
+ * @id ql/inline-overlay-caller
+ * @tags performance
+ * @precision high
+ */
+
+import ql
+
+predicate mayBeLocal(AstNode n) {
+  n.getAnAnnotation() instanceof OverlayLocal
+  or
+  n.getAnAnnotation() instanceof OverlayLocalQ
+  or
+  // The tree-sitter-ql grammar doesn't handle annotations on file-level
+  // module declarations correctly. To work around that, we consider any
+  // node in a file that contains an overlay[local] or overlay[local?]
+  // annotation to be potentially local.
+  exists(AstNode m |
```

Comment:
This should be good enough for now (given we insert `overlay[local?]` at the top of files), but I would imagine as QL writers take ownership of the annotations this will lead to annoying false-positives. I suppose we can fix it when it actually becomes a problem.

Reply:
Yes, I believe this will be good enough for now. I've filed https://github.com/github/codeql-core/issues/4953 to fix the tree-sitter grammar which is a prerequisite. 

Also, note that even with better attribution of overlay annotations on declarations, the query will still be based on an approximate locality analysis, unless we reimplement `QLOverlayLocalDetector` in QL.